File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/ProcessorMetrics.java
/*
 * Copyright (c) 2017-2018 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.redfish.base.DiscoverableEntity;
import com.intel.podm.business.entities.redfish.base.Entity;
import com.intel.podm.common.types.Id;

import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.OneToOne;
import javax.persistence.OrderColumn;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@Table(name = "processor_metrics", indexes = @Index(name = "idx_processor_metrics_entity_id", columnList = "entity_id", unique = true))
@SuppressWarnings({"checkstyle:MethodCount"})
public class ProcessorMetrics extends DiscoverableEntity {
    @Column(name = "entity_id", columnDefinition = ENTITY_ID_STRING_COLUMN_DEFINITION)
    private Id entityId;

    @Column(name = "bandwidth_percent")
    private BigDecimal bandwidthPercent;

    @Column(name = "average_frequency_mhz")
    private Integer averageFrequencyMhz;

    @Column(name = "throttling_celsius")
    private Integer throttlingCelsius;

    @Column(name = "temperature_celsius")
    private Integer temperatureCelsius;

    @Column(name = "consumed_power_watt")
    private Integer consumedPowerWatt;

    @ElementCollection
    @CollectionTable(name = "processor_metrics_health", joinColumns = @JoinColumn(name = "processor_metrics_id"))
    @Column(name = "health")
    @OrderColumn(name = "health_order")
    private List<String> healthList = new ArrayList<>();

    @OneToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "processor_id")
    private Processor processor;

    @Override
    public Id getId() {
        return entityId;
    }

    @Override
    public void setId(Id id) {
        entityId = id;
    }

    public BigDecimal getBandwidthPercent() {
        return bandwidthPercent;
    }

    public void setBandwidthPercent(BigDecimal bandwidthPercent) {
        this.bandwidthPercent = bandwidthPercent;
    }

    public Integer getAverageFrequencyMhz() {
        return averageFrequencyMhz;
    }

    public void setAverageFrequencyMhz(Integer averageFrequencyMhz) {
        this.averageFrequencyMhz = averageFrequencyMhz;
    }

    public Integer getThrottlingCelsius() {
        return throttlingCelsius;
    }

    public void setThrottlingCelsius(Integer throttlingCelsius) {
        this.throttlingCelsius = throttlingCelsius;
    }

    public Integer getTemperatureCelsius() {
        return temperatureCelsius;
    }

    public void setTemperatureCelsius(Integer temperatureCelsius) {
        this.temperatureCelsius = temperatureCelsius;
    }

    public Integer getConsumedPowerWatt() {
        return consumedPowerWatt;
    }

    public void setConsumedPowerWatt(Integer consumedPowerWatt) {
        this.consumedPowerWatt = consumedPowerWatt;
    }

    public List<String> getHealthList() {
        return healthList;
    }

    public void addHealth(String health) {
        this.healthList.add(health);
    }

    public Processor getProcessor() {
        return processor;
    }

    public void setProcessor(Processor processor) {
        if (!Objects.equals(this.processor, processor)) {
            unlinkProcessor(this.processor);
            this.processor = processor;
            if (processor != null && !this.equals(processor.getProcessorMetrics())) {
                processor.setProcessorMetrics(this);
            }
        }
    }

    public void unlinkProcessor(Processor processor) {
        if (Objects.equals(this.processor, processor)) {
            this.processor = null;
            if (processor != null) {
                processor.unlinkProcessorMetrics(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkProcessor(processor);
    }

    @Override
    public boolean containedBy(Entity possibleParent) {
        return isContainedBy(possibleParent, processor);
    }
}

File: PSME/application/src/rest/endpoints/system/processor_metrics.cpp
/*!
 * @brief ProcessorMetrics endpoint implementation
 *
 * @copyright Copyright (c) 2017-2019 Intel Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file processor_metrics.cpp
 */

#include "psme/rest/endpoints/system/processor_metrics.hpp"
#include "psme/rest/endpoints/utils.hpp"



using namespace psme::rest;
using namespace psme::rest::constants;

namespace {
json::Json make_prototype() {
    json::Json r(json::Json::value_t::object);

    r[Common::ODATA_CONTEXT] = "/redfish/v1/$metadata#Intel_RackScale.ProcessorMetrics.ProcessorMetrics";
    r[Common::ODATA_ID] = json::Json::value_t::null;
    r[Common::ODATA_TYPE] = "#Intel_RackScale.ProcessorMetrics.v1_0_0.ProcessorMetrics";
    r[Common::ID] = json::Json::value_t::null;
    r[Common::NAME] = "Processor Metrics";
    r[Common::DESCRIPTION] = "Processor Metrics";
    r[ProcessorMetrics::AVERAGE_FREQUENCY_MHZ] = json::Json::value_t::null;
    r[ProcessorMetrics::BANDWIDTH_PERCENT] = json::Json::value_t::null;
    r[ProcessorMetrics::CONSUMED_POWER_WATT] = json::Json::value_t::null;
    r[ProcessorMetrics::HEALTH] = json::Json::value_t::array;
    r[ProcessorMetrics::TEMPERATURE_CELSIUS] = json::Json::value_t::null;
    r[ProcessorMetrics::THROTTLING_CELSIUS] = json::Json::value_t::null;

    return r;
}

}


endpoint::ProcessorMetrics::ProcessorMetrics(const std::string& path) : EndpointBase(path) {}


endpoint::ProcessorMetrics::~ProcessorMetrics() {}


void endpoint::ProcessorMetrics::get(const server::Request& req, server::Response& res) {

    auto processor_uuid = psme::rest::model::find<agent_framework::model::System, agent_framework::model::Processor>(
        req.params).get_uuid();

    auto json = make_prototype();
    json[Common::ODATA_ID] = PathBuilder(req).build();
    json[Common::ID] = constants::Common::METRICS;
    json[Common::NAME] = "Processor " + req.params[PathParam::PROCESSOR_ID] + " Metrics";

    utils::populate_metrics(json, processor_uuid);

    set_response(res, json);
}
